src/main/java/de/zalando/ep/zalenium/container/swarm/SwarmContainerClient.java

```diff
@@ -0,0 +1,477 @@
+        try {
+            Task task = SwarmUtilities.getTaskByContainerId(containerId);
+            if (task != null) {
+                pullSwarmExecImage();
+                startSwarmExecContainer(task, command, containerId);
+            } else {
+                logger.warn("Couldn't execute command on container {}", containerId);
+            }
+        } catch (DockerException | InterruptedException e) {
+            logger.warn("Error while executing comman on container {}", containerId);
+            ga.trackException(e);
+        }
+    }
+
+    private void pullSwarmExecImage() {
+        try {
+            SwarmUtilities.pullImageIfNotPresent(SWARM_EXEC_IMAGE);
+        } catch (DockerException | InterruptedException e) {
+            logger.warn(nodeId + " Error while checking (and pulling) if the image is present", e);
+            ga.trackException(e);
+        }
+    }
+
+    private void startSwarmExecContainer(Task task, String[] command, String containerId) throws DockerException, InterruptedException {
+        String taskId = task.id();
+
+        List<String> binds = new ArrayList<>();
+        binds.add("/var/run/docker.sock:/var/run/docker.sock");
+
+        HostConfig.Builder hostConfigBuilder = HostConfig.builder()
+                .autoRemove(true)
+                .appendBinds(binds);
+
+        HostConfig hostConfig = hostConfigBuilder.build();
+
+        command[0] = "task-exec";
+        command[1] = taskId;
+
+        logger.debug("Executing command: {} - on Container: {}",
+                Arrays.toString(command),
+                containerId);
+
+        ContainerConfig containerConfig = ContainerConfig.builder()
+                .image(SWARM_EXEC_IMAGE)
+                .hostConfig(hostConfig)
+                .cmd(command)
+                .build();
+
+        SwarmUtilities.startContainer(containerConfig);
+    }
+
+    public String getLatestDownloadedImage(String imageName) {
+        // TODO: verify this is handled by docker
+        return imageName;
+    }
+
+    public ContainerCreationStatus createContainer(String zaleniumContainerName, String image, Map<String, String> envVars,
+                                                   String nodePort) {
+        // TODO: is it meaningful to add labels to identify services/containers in the swarm?
+
+        List<String> flattenedEnvVars = envVars.entrySet().stream()
+                .map(e -> e.getKey() + "=" + e.getValue())
+                .collect(Collectors.toList());
+        final ContainerSpec containerSpec = buildContainerSpec(flattenedEnvVars, image);
+
+        final TaskSpec taskSpec = buildTaskSpec(containerSpec);
+
+        String noVncPort = envVars.get("NOVNC_PORT");
+        final ServiceSpec serviceSpec = buildServiceSpec(taskSpec, nodePort, noVncPort);
+
+        try {
+            ServiceCreateResponse service = SwarmUtilities.createService(serviceSpec);
```

Comment:
I think it should be better if we pre-define the worker services, then update theirs scaling like `https://docs.docker.com/engine/swarm/swarm-tutorial/scale-service/`

Reply:
Hi @trinhpham,

Service scaling was our first Approach for creating and deleting Browser tasks. 

The problem was: When scaling down the number of replicas, it doesn't took in account which task is  not active. It killed containers that were actually running tests and made the tests fail. We found no way to specify the task to shutdown, when scaling down the number of replicas. 

https://github.com/moby/moby/issues/33426

If we scale the number of replicas up instead of creating new services. We will have to scale it down when the tests end. And this will lead to a wrong behaviour.

Reply:
Got your idea. Thanks for explaining it to me :)